Add SOP0012 for Withdraw Dataset from Node Allele Frequency Beacon (#73)
GenomicDataInfrastructure/standard-operating-procedures: sops/node-specific/GDI-SOP0012_withdraw_dataset_from_node_allele_frequency_beacon.md

> # European GDI - Withdraw_dataset_from_node_allele_frequency_beacon

Reviewer: fix title to _European GDI - Dataset Withdrawal from Node Allele Frequency Beacon_

Author: The title is already correct as it has to be written with underscores to follow the PR requirements.

> # European GDI - Withdraw_dataset_from_node_allele_frequency_beacon

Reviewer 2: If you inspect the linting code [here](https://github.com/GenomicDataInfrastructure/standard-operating-procedures/blob/c1198a3ba433e12b4ee7257bf87793b75af546fe/scripts/sop_linter.py#L623-L652), it takes away whitespaces and puts both the title and what comes after ``...GDI - `` here, and checks that they match.

Also, in the ISM document there's information on how to name the files. But in summary, what you should do is:
- Change this file's name, using ``-`` as separators after the ``GDI-SOP0012_`` underscore (i.e., ``withdraw-dataset-...``
- Change the title to:
```suggestion
# European GDI - Withdraw dataset from node allele frequency beacon
```
Or similar (lowercase/Uppercase is ignored)

| HTTP          | Hypertext Transfer Protocol                       |
| ID            | Identifier                                        |
| SOP           | Standard Operating Procedure                      |
| VHD           | Virtual Helpdesk                                  |

| Term          | Definition                                                                                                |
| ------------- | ----------------------------------------------------------------------------------------------------------|
| Hard-deletion | Complete data removal from primary systems, plus documented handling of backups according to retention policy, with escalation if the request requires something stricter.                                     |
| Permissions   | All the information related to the dataset grants and its security level configuration.                   |
| Query         | HTTP request to an endpoint of the node’s Allele Frequency beacon.                                        |
| Soft-deletion | Data is marked as withdrawn and made inaccessible to users but retained internally for audit or limited-term retention.|

> | Withdrawal    | A performance action to remove the dataset from beacon responses.                                         |

Reviewer: maybe replace performance with administrative or operational

Author: Now using the greater level charter definition to have all definition aligned. Thanks!

| Role       | Full name                   | GDI/node role                                    | Organisation                          |
| ---------- | --------------------------- | ------------------------------------------------ | ------------------------------------- |
| Author     | Liina Nagirnaja             | Beacon Manager                                   | CRG                                   |
| Author     | Jordi Rambla                | Beacon Product Owner                             | CRG                                   |
| Author     | Oriol López-Doriga Sagalés  | Beacon Developer                                 | CRG                                   |
| Reviewer   | Marcos Casado Barbero       | Task 4.3 member                                  | EMBL-EBI                              |
| Approver   | Gabriele Rinck              | Task 4.3 member                                  | EMBL-EBI                              |
| Authorizer | Management Board            | Authorizer according to GDI SOP governance       | GDI                                   |

### 4. Purpose

> AF Beacons expose population-level allele counts. A node-level SOP to withdraw data from Allele Frequency Beacons guarantees consistent, auditable removal of frequency contributions in line with GDPR and the GDI overarching withdrawal workflow.

Reviewer: Small wording suggestion for clarity and flow:
"AF Beacons provide access to population-level allele counts. This node-level SOP defines the process for withdrawing data from Allele Frequency Beacons, ensuring consistent and auditable removal of these contributions in line with GDPR and the overarching GDI withdrawal workflow."

> AF Beacons expose population-level allele counts. A node-level SOP to withdraw data from Allele Frequency Beacons guarantees consistent, auditable removal of frequency contributions in line with GDPR and the GDI overarching withdrawal workflow.

Reviewer 2: ```suggestion
Allele Frequency (AF) Beacons expose population-level allele counts. A node-level SOP to withdraw data from Allele Frequency Beacons guarantees consistent, auditable removal of frequency contributions in line with GDPR and the GDI overarching withdrawal workflow.
```

Author: I'm adding AF to glossary, AF Beacons should be correct IMO and no need to accept this change.

Reviewer 2: But isn't it common to expand the acronyms the first time they are used? That's why I was expanding it here, but if there's another expansion above, no need then

### 5. Scope

This SOP covers the steps required to withdraw a dataset from a Node's Allele Frequency Beacon. Included steps are the identification of dataset records to withdraw, assessing the withdrawal scope, executing the removal of data, validating frequency integrity, auditing and logging the withdrawal, and reporting completion.
It is part of the higher level Dataset Withdrawal SOP (SOP0009).

### 6. Introduction and Background Information

> An aggregated data beacon hosts information for variants and minimal metadata for datasets that these variants belong to. These two types of different information are often stored separately in a database but linked through some reference property that contains the dataset id for the records. This dataset id is the pivotal piece of information, under which the records are displayed in a client response and that allows to integrate the response with other services in GDI.

Reviewer: - The term "aggregated data beacon" is introduced here, while earlier the document refers to "AF Beacon" / "Allele Frequency Beacon"
- why "minimal" metadata for dataset? is this correct?
- maybe skip "different" in "_two types of different information_"
- Abbreviation  defined in the table is ID, you are using id.

Reviewer 2: Agreed

> An aggregated data beacon hosts information for variants and minimal metadata for datasets that these variants belong to. These two types of different information are often stored separately in a database but linked through some reference property that contains the dataset id for the records. This dataset id is the pivotal piece of information, under which the records are displayed in a client response and that allows to integrate the response with other services in GDI.

Reviewer 2: Are records _owned_ by a dataset entity? Or just _linked_ by it? I'm asking in case a dataset can be withdrawn but their record be reused in a different dataset. 

Similar to the case of CASCADE in SQL deletions, if that helps.

Author: The records are owned and linked by a dataset entity. It's not that you can't update a record to change the dataset a record belongs to, but is not something I expect (recommend) to happen because everything can get messy. Yes, the CASCADE in SQL is similar to what we are elaborating for the hard-deletion dataset withdrawal.

> When attempting a withdrawal of the beacon records related to a dataset, it will be necessary to determine which is the exact id of the dataset, as case sensitivity and other character inferences may result in an unsuccessful withdrawal attempt. The withdrawal procedure of a soft-deletion is slightly different from a hard-deletion, where the amount of records to be deleted is significantly higher. Withdrawal by hard-deletion includes variants and dataset records type, whereas soft-deletion includes only the latter. 

Reviewer 2: ```suggestion
When attempting a withdrawal of the beacon records related to a dataset, it will be necessary to determine which is the exact id of the dataset, as case sensitivity and other character inferences may result in an unsuccessful withdrawal attempt. The withdrawal procedure of a soft-deletion is slightly different from a hard-deletion, where the amount of records to be deleted is significantly higher. Withdrawal by hard-deletion includes variant and dataset record types, whereas soft-deletion includes only the latter. 
```

Author: Thanks, I  had to add this manually as I already committed this file and I had conflicts.

    B -->|Soft-deletion| C[Remove dataset record]
    B -->|Hard-deletion| D[Remove dataset and variants records]
    C --> E[Remove permissions for the dataset]
    D --> E
    E --> F[Log action]
    F --> G[Report completion to VHD]
```

### 8. Procedure

#### 8.1. Confirm withdrawal request and identify dataset

| Step identifier | When | Who |
| :-------------- | :------------------------------------------------------------------ | :-------------------------------------- |
| `1`             | When a dataset withdrawal request has been approved under the overarching withdrawal process described in [GDI-SOP0009 dataset withdrawal](../european-level/GDI-SOP0009_dataset-withdrawal.md#87-per-system-dataset-withdrawal). | Node aggregated beacon maintainer |

As the node aggregated beacon maintainer, confirm that the incoming request package is complete before performing any deletion in beacon. Double check first:
- The original approved withdrawal request or ticket reference
- The requested scope of the withdrawal (soft or hard-deletion)
- The relevant dataset identifier matching the node FAIR Data Point’s dataset to be withdrawn

> Send an HTTP GET request to your beacon’s datasets endpoint and locate the record corresponding to the ID of the dataset to be deleted,

Reviewer: move this to new line

#### 8.2. Remove the dataset from the beacon database

| Step identifier | When | Who |
| :-------------- | :------------------------------------------------------------------ | :-------------------------------------- |
| `2`             | After successful completion of [Step 1](#81-confirm-withdrawal-request-and-identify-dataset). | Node aggregated beacon maintainer |

After deletion type of the dataset withdrawal is confirmed:
- If the deletion type is a soft-deletion, proceed to ⏩[Step 2.1](#821-soft-deletion).
- If the deletion type is a hard-deletion, proceed to ⏩[Step 2.2](#822-hard-deletion-option-a).

##### 8.2.1. Soft-deletion

| Step identifier | When | Who |
| :-------------- | :--------------------------------------- | :-------------------------------------- |
| `2.1`             | The deletion type was identified as soft. | Node aggregated beacon maintainer |

Locate yourself in the server where your beacon is running and execute the following command line script to remove the dataset record approved for  withdrawal.
```bash

> docker exec mongoprod /bin/bash -c 'mongosh beacon -u <user> -p <password> --authenticationDatabase admin --eval "db.datasets.deleteMany({\"id\": \"<id>\"})"'

Reviewer 2: Shouldn't we have a step before this one, where the maintainer checks in the server that the dataset is there?

Mainly for verification, to see that it appeared before, and that it doesn't appear later. Mainly asking in case the query ``curl 'https://<yourBeaconDomain>/datasets/<id>'`` may have a different result than the server (e.g., imagine someone gets into the wrong server or is not sure).

Reviewer 2: Also, for a similar reason, is there a way for the maintainer to assert that they are in the correct server? For example, that they are not in a ``dev`` or mirrored server, and that the change will be propagated to production.

Author: Hmm, I think this would be like trying to over-control all the user errors, and if we start like this it will be a never ending story. Sorry but not agree on this. The user must be aware of their instances and when is he working with a `prod` or a `dev` version and in `prod` be as cautious as possible.

Reviewer 2: Fair enough - I come from the point of view where I try to minimise errors, as it's the aim of an SOP. If it's an overkill, disregard my comment 👍

- If the removal is not successful, record the response obtained from the used commands, adding all the information about the actions performed and the intended goal of performing them and report to the GDI Virtual Helpdesk so that requester communication continues through the VHD workflow and proceed to ⏩[Step 3](#83-remove-dataset-permissions).

#### 8.3. Remove dataset permissions

| Step identifier | When | Who |
| :-------------- | :--------------------------------------- | :-------------------------------------- |
| `3`             | After ⏩[Step 2](#82-remove-the-dataset-from-the-beacon-database). | Node aggregated beacon maintainer |

Located at the server where your beacon is running, proceed to edit the file at the path `/beacon/permissions/datasets/datasets_permissions.yml` and remove the dataset entry.
Save the file.
- Continue to ⏩[Step 4](#84-verify-log-and-report-completion).

#### 8.4. Verify, log, and report completion

| Step identifier | When | Who |
| :-------------- | :--------------------------------------- | :-------------------------------------- |
| `4`             | After successful completion of Step 3 | Node aggregated beacon maintainer |

> Verify in the beacon that the withdrawn records in the dataset are not appearing in a query of datasets and a query of variants.

Reviewer: Maybe add curl commands, for example:
```bash
curl 'https://<yourBeaconDomain>/datasets/<id>'
curl 'https://<yourBeaconDomain>/datasets/<id>/g_variants'
```

Author: Changed this, now not needed anymore. Thanks!

> Add the dataset id to the file in path `/beacon/conf/datasets/datasets/datasets_conf.yml` and add a new item under it called isDeprecated setting it as True:

Reviewer: The path "/beacon/conf/datasets/datasets/datasets_conf.yml" seems incorrect — there is an extra "datasets" in the path. Should one of them be removed?

#### 8.2. Remove the dataset from the beacon database

| Step identifier | When | Who |
| :-------------- | :------------------------------------------------------------------ | :-------------------------------------- |
| `2`             | After successful completion of [Step 1](#81-confirm-withdrawal-request-and-identify-dataset). | Node aggregated beacon maintainer |

After deletion type of the dataset withdrawal is confirmed:
- If the deletion type is a soft-deletion, proceed to ⏩[Step 2.1](#821-soft-deletion).
- If the deletion type is a hard-deletion, proceed to ⏩[Step 2.2](#822-hard-deletion-option-a).

##### 8.2.1. Soft-deletion

| Step identifier | When | Who |
| :-------------- | :--------------------------------------- | :-------------------------------------- |
| `2.1`             | The deletion type was identified as soft. | Node aggregated beacon maintainer |

Locate yourself in the server where your beacon is running and execute the following command line script to remove the dataset record approved for  withdrawal.
```bash

> docker exec mongoprod /bin/bash -c 'mongosh beacon -u <user> -p <password> --authenticationDatabase admin --eval "db.datasets.deleteMany({\"id\": \"<id>\"})"'

Reviewer 2: As a rule of thumb I wouldn't recommend passing user and password in the shell command, but have a ``.netrc`` or ``.mongoshrc`` file referenced. Otherwise, the history may leak credentials.

Author: Ok, I didn't use a file because it then has to be mounted to the container and we don't have this mechanics ready for the latest release. I added a script to prompt the credentials on script runtime and never leaked in shell history.

Reviewer 2: That's alright then. As I said... I've never maintained a Beacon myself, so some of my comments may not be that necessary
